pgjdbc/src/main/java/org/postgresql/largeobject/BlobOutputStream.java

```diff
@@ -80,36 +99,44 @@ public void write(byte[] b, int off, int len) throws java.io.IOException {
     try (ResourceLock ignore = lock.obtain()) {
       LargeObject lo = checkClosed();
       byte[] buf = this.buf;
-      // Initialize the buffer if needed
-      // Suppose we have bufferPosition=10KiB buffered out of bufferSize=64KiB, and the user issues
-      // 100KiB write. We will copy 54KiB to the buffer and flush it, then we will copy
-      // the remaining 46KiB to the buffer without flushing to the DB. It avoids small writes
-      // hitting the database.
-      // If the incoming request is 300KiB, then we copy 54KiB to the buffer and flush it.
-      // Then we write the remaining 246KiB directly to the database.
-      // At worst, it results in two DB calls
-      // If the buffer was not there, we do not create one if the write request is big enough
-      if (buf == null && len < 2 * bufferSize) {
-        this.buf = buf = new byte[bufferSize];
-      }
-      if (buf != null && bufferPosition < bufferSize) {
-        // Copy the part of the user-provided data that fits in the buffer
-        int avail = Math.min(bufferSize - bufferPosition, len);
-        System.arraycopy(b, off, buf, bufferPosition, avail);
-        bufferPosition += avail;
-        len -= avail;
-        off += avail;
-      }
-      if (len == 0) {
-        return;
+      int totalData = bufferPosition + len;
+      // We want aligned writes, so the write requests chunk nicely into large object rows
+      int tailLength =
+          bufferSize >= 8192 ? totalData % 8192 : (
```

Comment:
tail is really remainder, correct ? I'm having trouble coming up with a better name, but can we document it ?

Reply:
That is indeed a remainder. Will probably add ACSII art comment

Reply:
Thanks! I'm just acknowledging that in 6 months I won't remember this discussion :(